akka-http-core/src/main/scala/akka/http/impl/util/SocketOptionSettings.scala

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.http.impl.util
+
+import akka.io.{ Tcp, Inet }
+
+import scala.collection.immutable
+
+import akka.io.Inet.SocketOption
+import com.typesafe.config.Config
+
+private[http] object SocketOptionSettings {
+  def fromSubConfig(config: Config): immutable.Traversable[SocketOption] = {
+    def so[T](setting: String)(f: (Config, String) ⇒ T)(cons: T ⇒ SocketOption): List[SocketOption] =
+      config.getString(setting) match {
+        case "undefined" ⇒ Nil
+        case x           ⇒ cons(f(config, setting)) :: Nil
+      }
+
+    (so("so-receive-buffer-size")(_ getIntBytes _)(Inet.SO.ReceiveBufferSize) :::
+      so("so-send-buffer-size")(_ getIntBytes _)(Inet.SO.SendBufferSize) :::
+      so("so-reuse-address")(_ getBoolean _)(Inet.SO.ReuseAddress) :::
+      so("so-traffic-class")(_ getInt _)(Inet.SO.TrafficClass) :::
+      so("tcp-keep-alive")(_ getBoolean _)(Tcp.SO.KeepAlive) :::
+      so("tcp-oob-inline")(_ getBoolean _)(Tcp.SO.OOBInline) :::
+      so("tcp-no-delay")(_ getBoolean _)(Tcp.SO.TcpNoDelay)).toList
+  }
```

Comment:
This would mean:
 * building Options for each of the elements
 * putting these into an intermediate sequence to accommodate the `List.apply` varargs constructor
 * convert this intermediate sequence to a list of options
 * run flatten which will need to unpack all elements again to build another (though, probably empty) list

The existing code will create only Cons elements for defined elements with the default case of all being `Nil`, so that no intermediate structures need to be built at all. The same happens when only one of the settings is set. With more than one element set, `:::` will use a `ListBuffer` per step to append two lists. So maybe we should use `reverse_:::` which would be even more efficient but also way more ugly, so I'd rather keep it at that :).